### ISSUE TITLE
Improve documentation

### ISSUE THREAD

**Issue by edvin** [edvin/tornadofx#56]
Some of the documentation pages are becoming rather lengthy and might be hard to follow, like for example the [Type Safe Builders page](https://github.com/edvin/tornadofx/wiki/Type-Safe-Builders). We need to think about a better structure, and possibly a better format for the documentation.

Some options:

1. Restructure/split the wiki pages
2. Create a dedicated web page for the project with documentation, maybe something like the [JacpFX Documentation](http://jacpfx.org/documentation_main.html)
3. Start thinking about/planning a book
4. Create a larger sample application that touch as much of the functionality as possible. This could also possibly be the basis for a book.

**Comment by thomasnield** [comment]
I've been thinking about this too. Perhaps we can create an eBook GitHub project like this one for RxJava?

https://github.com/Froussios/Intro-To-RxJava

It's unlikely we will get anything published unless we self-publish. If I do write a "Kotlin for Beginners" book in a year or two, it is my plan to dedicate a chapter to this library.

**Comment by edvin** [comment]
eBook GitHub might be a good idea, easy to collaborate etc. Yeah, you should definitely do a "Kotlin for Beginners" in time :) https://www.bookbaby.com seems to be a good place to self publish. I'd get my company to sponsor all expenses. I don't think we'd make much money on writing a book now, but writing in book form forces us to structure the information a bit more meticulously than a couple of wiki pages.

I'll start looking into a web page for the project in the mean time.

**Comment by thomasnield** [comment]
Okay awesome, glad to hear your company can help sponsor things when we need to scale to more formal documentation. I vote we try the GitHub book in the interim, because that Intro to RxJava book seems to set a good example and as you say, it is easy to collaborate. A website might be good once documentation slows down.

**Comment by edvin** [comment]
I think you're right. For now, we'll keep the documentation in the Wiki, and then we could start working on something in book form in the GitHub repo and later convert it to an actual book.

I'll hold off on the website until documentation and features slows down :) Closing for now, was good to just discuss it quickly :)

**Comment by thomasnield** [comment]
Awesome! Let me know when you think its time to create the GitHub "book" project. I'll be sure to contribute to it when its time...

**Comment by ronsmits** [comment]
Once you start documenting, please spend some time on the @FXML lateinit construct. I find it hard to get it to work decently and find myself working around it.

**Comment by edvin** [comment]
The *@FXML* lateinit construct should no longer be needed, I'll be sure to update the documentation where it is still mentioned. This is the preferred way now:

```kotlin
override val root: VBox by fxml()
val myButton: Button by fxid()
```

No need to use the @FXML annotation, you get a `val` instead of a `var` and you avoid that pesky `lateinit` construct.

Would you care to elaborate on the problem you have run into when using @FXML though? I would like to see if there is anything I can do to make this more robust, it is supported after all.

**Comment by edvin** [comment]
@ronsmits have you got a minute to comment on the @FXML issue you're having? See my answer above as well :)

**Comment by ronsmits** [comment]
Sorry for not seeing this sooner, I will setup a use case tonight and make it a small project

**Comment by edvin** [comment]
Perfect :)

**Comment by ronsmits** [comment]
I created a very very small example https://github.com/ronsmits/example-tornadofx. With 1.4.0 and 1.4.1 the 
````
@FXML lateinit var headerText : Label
````
works ````var headerText : Label by fxml()```` does not even compile.

**Comment by edvin** [comment]
OK, I'll have a look right away. The reason your second example doesn't compile is that you're using the `fxml()` instead of the `fxid()` delegate :)

In short: `fxml()` resolves a Node by loading an FXML document, and `fxid()` resolves a node in the loaded FXML document by it's `fx:id`.

**Comment by ronsmits** [comment]
I did not push the not compiling code, so I just did that for you

**Comment by edvin** [comment]
When you use `fxid()` you must define your properties as immutable, so this will work:

```kotlin
val otherText: Label by fxid()
```

Just confirmed it by pulling the latest commit and changing var to val :)

**Comment by edvin** [comment]
Also, your example seems to work fine for me, I get a white window with the text "dit werkt niet".

When you run it, what error message/stack trace do you get?

**Comment by ronsmits** [comment]
my comments were not clear, the example I made worked with 1.4.0 and 1.4.1. With the val otherText that works too. So currently I cannot reproduce the issues I had.

**Comment by edvin** [comment]
Ah, OK :) Please let me know if you notice anything off with the FXML support later on :)
